src/haz3lcore/statics/Statics.re

```diff
@@ -246,9 +247,22 @@ and uexp_to_info_map =
     );
   | Let(p, def, body) =>
     let (p_syn, _) = go_pat(~is_synswitch=true, ~mode=Syn, p, m);
-    let def_ctx = extend_let_def_ctx(ctx, p, p_syn.ctx, def);
-    let (def, m) = go'(~ctx=def_ctx, ~mode=Ana(p_syn.ty), def, m) /* Analyze pattern to incorporate def type into ctx */;
-    let (p_ana, m) = go_pat(~is_synswitch=false, ~mode=Ana(def.ty), p, m);
+    let (def, p_ana, m) =
+      if (Option.is_none(UExp.get_recursive_bindings(p, def))) {
+        let (def, m) = go(~mode=Ana(p_syn.ty), def, m) /* Analyze pattern to incorporate def type into ctx */;
+        let (p_ana, m) =
+          go_pat(~is_synswitch=false, ~mode=Ana(def.ty), p, m);
+        (def, p_ana, m);
+      } else {
+        let (def_base, _) = go'(~ctx=p_syn.ctx, ~mode=Ana(p_syn.ty), def, m) /* Analyze pattern to incorporate def type into ctx */;
+        let (p_ana, m) =
+          go_pat(~is_synswitch=false, ~mode=Ana(def_base.ty), p, m);
+        let (def, m) = go'(~ctx=p_ana.ctx, ~mode=Ana(p_syn.ty), def, m);
```

Comment:
In the recursive case, as the part of the context associated with the pattern becomes more accurate, the context for the definition expression becomes outdated and requires updating.